### DISCUSSION TITLE
How to correctly divide the TUdataset class into training sets and test sets ?

### DISCUSSION THREAD

**Discussion by SongZan222** [pyg-team/pytorch_geometric#6432] [Q&A] [answered]
Dear PyG community,

For a graph neural network that classifies graphs as tasks, I mainly used data sets from TUDataset. But this class does not seem to provide parameters that divide into training sets and test sets.
I tried to manually divide the obtained dataset into training and test sets using the following code. In this case, the model works fine on the training set, but the accuracy on the test set is 0.
I have tried looking for instructions on dividing TUDataset on the Internet and in official documents but have found none.

![Figure_1](https://user-images.githubusercontent.com/93072497/212606970-de766329-c38a-4f9b-b18a-a45e12f09919.png)

```python
import numpy as np
import torch
import torch_geometric
from torch_geometric.datasets import TUDataset
from torch_geometric.loader import DataLoader
from torch_geometric.graphgym import register_pooling
from  torch_geometric.nn import GCNConv
import torch_scatter
import scipy
from torch_geometric.nn import aggr
import matplotlib.pyplot as plt
import networkx

# 0.0 Hyper-parameters 超参数
DEVICE = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
DATASET_NAME = 'ENZYMES'
LEARNING_RATE = 0.01
EPOCH = 128
BATCH_SIZE_TRAIN = 50
BATCH_SIZE_TEST = 10
# --------------------------------- #
# 0.1 Auxiliary function 辅助函数
# Draw a line graph  绘制折线图
def plt_line_graph(list_x,list_y1):
    '''

    :param list_x: The horizontal axis of the line graph，折线图的横轴
    :param list_y1: The first set of data that needs to be drawn 第一组数据
    :param list_y2: The second set of data that needs to be drawn 第二组数据
    :return: Draw the line graph
    '''
    plt.plot(
        list_x,
        list_y1,
        label="loss",
        color='#58006E'
    )
    # y1 Average value   绘制y1平均值
    y1_average = [sum(list_y1) / len(list_y1), sum(list_y1) / len(list_y1)]
    x_first_and_end = [list_x[0], list_x[-1]]
    plt.plot(
        x_first_and_end,
        y1_average,
        label="average loss",
        color='#E5439A',
        linestyle='-.',
        markersize=15
    )
    # plt.plot(
    #     list_x,
    #     list_y2,
    #     label="Accuracy Rate after N epochs",
    #     color='#F6573D',
    # )
    # #  y2 Average value 绘制y2平均值
    # y2_average = [sum(list_y2) / len(list_y2), sum(list_y2) / len(list_y2)]
    # x_first_and_end = [list_x[0], list_x[-1]]
    # plt.plot(
    #     x_first_and_end,
    #     y2_average,
    #     label="average accuracy",
    #     color='#FC9F81',
    #     linestyle='-.',
    #     markersize=15
    # )
    plt.legend(loc='best')
    # Control Y-axis range and step size 控制y轴范围和步长
    # plt.yticks(y_ticks[10:90:10])
    plt.grid(  # Background dotted line grid 背景虚线网格
        True,
        linestyle='--',
        alpha=0.25
    )
    # The name of the axis and the name of the chart 轴的名称和表的名称
    plt.xlabel("idx", fontdict={'size': 16})
    plt.ylabel("loss and accuracy", fontdict={'size': 16})
    plt.title("Test Result", fontdict={'size': 16})

    plt.show()

# Drawing matrix 绘制矩阵
def plt_matrix(m, name='the image of matrix'):
    m=plt.matshow(m, cmap=plt.cm.RdPu)
    plt.colorbar(m.colorbar, fraction=0.25)
    plt.title(name)
    plt.show()


# Plot the graph data structure 绘制图数据结构
def plt_dataset_sample(dataset, idx):
    '''

    :param dataset: The data set to display 要显示的数据集
    :param idx: Sample with serial number idx 绘制第idx个样本
    :return:Draw the adjacency matrix and the image 绘制邻接矩阵和图像
    '''
    data = dataset
    G = torch_geometric.utils.to_networkx(data[idx])
    print(type(G))
    networkx.draw_networkx(
        G=G,
        pos=networkx.spring_layout(G),
        node_size=150,
        cmap=plt.cm.RdPu,
        width=1,
        style='solid',
        label='A graph in the data set'
    )
    plt.show()



# 1. Prepare data set  准备数据集
dataset = TUDataset(
    root=r'C:\Users\Administrator\Desktop\data',
    name= DATASET_NAME
)
# Datalorder
def get_dataloader(name,is_train= True):
    dataset = TUDataset(
        root=r'C:\Users\Administrator\Desktop\data',
        name=name,
        transform=torch_geometric.transforms.ToDevice(DEVICE)
    )
    loader_train = DataLoader(
        dataset[0:400],
        shuffle=True,
        batch_size=BATCH_SIZE_TRAIN
    )
    loader_test = DataLoader(
        dataset[400:],
        shuffle=True,
        batch_size=BATCH_SIZE_TEST
    )
    if is_train:
        return loader_train
    else:
        return loader_test

# 2.Defined network 定义网络

sum_aggr =aggr.SumAggregation().to(DEVICE)
# Network definition
class Net(torch.nn.Module):
    def __init__(self):
        super(Net,self).__init__()
        self.conv1 = GCNConv(dataset.num_node_features,16)
        self.conv2 = GCNConv(16, 32)
        self.fc1 = torch.nn.Linear(32,16)
        self.fc2 = torch.nn.Linear(16,dataset.num_classes)

    def forward(self,data):
        x, edge_index = data.x, data.edge_index
        x = self.conv1(x,edge_index)
        x = torch.nn.functional.relu(x)
        x = self.conv2(x,edge_index)
        x = torch.nn.functional.relu(x)
        x = sum_aggr(x, ptr=data.ptr)
        x = self.fc1(x)
        x = self.fc2(x)
        return torch.nn.functional.log_softmax(x,dim=-1)

# Model instantiation 模型实例化
model = Net().to(DEVICE)
# optimizer
optimizer = torch.optim.Adam(model.parameters(),lr=LEARNING_RATE)


# 3.Train and Test
# train
def train(epoch):
    data_loder = get_dataloader(DATASET_NAME)
    counter = 0
    for idx in data_loder:
        output = model(idx).squeeze()
        optimizer.zero_grad()
        loss = torch.nn.functional.nll_loss(output,idx.y)
        loss.backward()
        optimizer.step()
        print(epoch,counter,loss.item())
        counter +=1

def test():
    model.eval()
    dataset = get_dataloader(DATASET_NAME, is_train=False)
    counter = 0
    test_counter = []
    test_acc_collecter = []
    for batch in dataset:
        out = model(batch).max(1)[1]
        test_acc = torch.eq(
            out, batch.y
        )
        test_counter.append(counter)
        counter += 1
        test_acc_collecter.append(test_acc.float().mean().item())
        print(batch.y)
        print(out)
        print(test_acc.float().mean())
    plt_line_graph(test_counter, test_acc_collecter)

for epoch in range(100):
    train(epoch)

test()

```

**Comment by rusty1s** [comment, marked as answer]
Yes, sadly `TUDataset` does not pre-define splits. One issue I see with your code is that you are not shuffling examples before doing the split. For a random split (which is usually used for these datasets), try
```python
dataset = dataset.shuffle()
train_dataset = dataset[:400]
test_dataset = dataset[400:]

**Reply by SongZan222** [reply]
The problem has been solved! rusty1s is my savior !
With rusty1s help, TUDataset can now be manually split into training sets and test sets.
```python

DEVICE = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
DATASET_NAME = 'DD'
BATCH_SIZE_TRAIN = 50
BATCH_SIZE_TEST = 10

def get_dataloader(name,is_train= True):
    dataset = TUDataset(
        root=r'C:\Users\Administrator\Desktop\data',
        name=name,
        transform=torch_geometric.transforms.ToDevice(DEVICE)
    ).shuffle()
    train_dataset = dataset[:400]
    test_dataset = dataset[400:]
    # The data set is shuffled, then sliced, and finally used as a Dataloader parameter
    # The range of data sets to be divided is set manually based on the size of each data set
    loader_train = DataLoader(
        train_dataset,
        shuffle=False,
        batch_size=BATCH_SIZE_TRAIN
    )
    loader_test = DataLoader(
        test_dataset,
        shuffle=False,
        batch_size=BATCH_SIZE_TEST
    )
    if is_train:
        return loader_train
    else:
        return loader_test

```
The model was tested on several data sets in the TUDataset class.
Here are the results.
![DD_test](https://user-images.githubusercontent.com/93072497/212621718-d5244e3f-965e-4c5f-9cc4-af3e6acbba5b.png)
![ENZYMES——test](https://user-images.githubusercontent.com/93072497/212621734-eed9a9c8-e0a4-4b5f-9baa-b497eec53dad.png)
![PROTEINS_test](https://user-images.githubusercontent.com/93072497/212621738-3eae03f8-37d7-49a7-96fe-06eb38f5793f.png)

Thank you for your help. It really helped me a lot with my paper.💕
